src/engine/strat_engine/devlinks.rs

```diff
@@ -0,0 +1,118 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+use std::collections::HashSet;
+use std::fs;
+use std::io::ErrorKind;
+use std::os::unix::fs::symlink;
+use std::path::{Path, PathBuf};
+use std::str;
+
+use super::super::errors::EngineResult;
+
+use engine::Engine;
+
+pub const DEV_PATH: &str = "/dev/stratis";
+
+/// Set up directories and symlinks under /dev/stratis based on current
+/// config. Clear out other stuff.
+// Don't just remove and recreate everything in case there are processes
+// (e.g. user shells) with the current working directory within the tree.
+pub fn setup_devlinks(engine: &Engine) -> EngineResult<()> {
+    if let Err(err) = fs::create_dir(DEV_PATH) {
+        if err.kind() != ErrorKind::AlreadyExists {
+            return Err(From::from(err));
+        }
+    }
+
+    let mut existing_dirs: HashSet<_> = fs::read_dir(DEV_PATH)?
+        .map(|dir_e| {
+                 dir_e
+                     .expect("direntry is valid")
+                     .file_name()
+                     .into_string()
+                     .expect("Unix is utf8")
+             })
+        .collect();
+
+    for pool in engine.pools() {
+        if !existing_dirs.contains(pool.name()) {
+            pool_added(pool.name());
```

Comment:
I would suggest that if you can't add the thing, you should just fail. Your method claims to do X. If it can't do X, and this is part of doing X, then returning an error is the correct thing to do, because it didn't do X.

Reply:
ok. changing everything to return a result instead of `()`.